src/wp-includes/class-wp-block.php

```diff
@@ -497,6 +461,51 @@ private function replace_html( string $block_content, string $attribute_name, $s
 		}
 	}
 
+	private static function get_block_bindings_processor( string $block_content ) {
+		$internal_processor_class = new class('', WP_HTML_Processor::CONSTRUCTOR_UNLOCK_CODE) extends WP_HTML_Processor {
+			/**
+			 * Replace the rich text content between a tag opener and matching closer.
+			 *
+			 * When stopped on a tag opener, replace the content enclosed by it and its
+			 * matching closer with the provided rich text.
+			 *
+			 * @param string $rich_text The rich text to replace the original content with.
+			 * @return bool True on success.
+			 */
+			public function replace_rich_text( $rich_text ) {
+				if ( $this->is_tag_closer() || ! $this->expects_closer() ) {
+					return false;
+				}
+
+				$depth = $this->get_current_depth();
+
+				$this->set_bookmark( '_wp_block_bindings_tag_opener' );
+				// The bookmark names are prefixed with `_` so the key below has an extra `_`.
+				$tag_opener = $this->bookmarks['__wp_block_bindings_tag_opener'];
+				$start      = $tag_opener->start + $tag_opener->length;
+				$this->release_bookmark( '_wp_block_bindings_tag_opener' );
+
+				// Find matching tag closer.
+				while ( $this->next_token() && $this->get_current_depth() >= $depth ) {
+				}
+
+				$this->set_bookmark( '_wp_block_bindings_tag_closer' );
+				$tag_closer  = $this->bookmarks['__wp_block_bindings_tag_closer'];
+				$end         = $tag_closer->start;
+				$this->release_bookmark( '_wp_block_bindings_tag_closer' );
+
+				$this->lexical_updates[] = new WP_HTML_Text_Replacement(
+					$start,
+					$end - $start,
+					$rich_text
+				);
```

Comment:
applying lexical updates that span token boundaries is not something we have particularly tested well. it’s probably working well enough and hopefully the tests here will catch them, but this is sort of uncharted territory within the HTML Processor. we should make sure that we don’t seek anywhere here to verify that we don’t cross indices.

this single issue is one of the main reasons I proposed the single linear pass of the serialization builder, because it structurally prevents the ability of jumping around or referring to indices after changing positions.

I’m not particularly confident I would know how this interacts with virtual tokens, for example, whose boundaries are coincident with bookmarks (vs. realized tokens which all have non-overlapping spans)